### PR TITLE
Miri: fix determining size of an "extra function" allocation

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes [a bug](https://github.com/rust-lang/miri/pull/862) introduced by https://github.com/rust-lang/rust/pull/62982. Best reviewed commit-by-commit.

r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
The function now locks the global map twice for gloabal non function pointer alloc IDs. Probably irrelevant perf wise though.

**Comment by RalfJung** [comment]
> The function now locks the global map twice for gloabal non function pointer alloc IDs. Probably irrelevant perf wise though.

Indeed. It will only do the fn ptr lookup though for allocations that are not in the local `alloc_map`, so only for a small fraction. I'd offer to perf both versions but you seem to be fine with this since you already r+'d?

**Comment by oli-obk** [comment]
I'm fine with it. We had that before my static cycle PR and I didn't see a perf improvement on that PR, so...

File: src/librustc_mir/interpret/machine.rs [hunk 1]
```diff
@@ -54,6 +54,22 @@ pub trait AllocMap<K: Hash + Eq, V> {
         k: K,
         vacant: impl FnOnce() -> Result<V, E>
     ) -> Result<&mut V, E>;
+
+    /// Read-only lookup.
+    fn get(&self, k: K) -> Option<&V> {
+        match self.get_or(k, || Err(())) {
```
Comment (Centril) [inline src/librustc_mir/interpret/machine.rs]:
```suggestion
        self.get_or(k, || Err(())).ok()
```
(and below)

Comment (RalfJung) [reply src/librustc_mir/interpret/machine.rs]: Ah, so much better :)

**Pull request merged**